lib/pengine/bundle.c

```diff
@@ -1269,14 +1278,16 @@ pe__unpack_bundle(pe_resource_t *rsc, pe_working_set_t *data_set)
                 pe__set_resource_flags(bundle_data->child, pe_rsc_notify);
             }
 
-            offset += allocate_ip(bundle_data, replica, buffer+offset,
-                                  max-offset);
+            allocate_ip(bundle_data, replica, buffer);
             bundle_data->replicas = g_list_append(bundle_data->replicas,
                                                   replica);
             bundle_data->attribute_target = g_hash_table_lookup(replica->child->meta,
                                                                 XML_RSC_ATTR_TARGET);
         }
-        bundle_data->container_host_options = buffer;
+        bundle_data->container_host_options = strdup((const char *)
+                                                     buffer->str);
```

Comment:
Good point.
It's questionable whether we need to strdup before assigning to a char * anyway. I've been doing it out of an abundance of caution for memory management (g_free vs. free). But there are existing places in the code where a gstring->str is assigned directly to a char pointer and then g_string_free(gstring, FALSE) is called (freeing the gstring but not gstring->str).

Reply:
As an FYI, since I got curious... Since glib 2.46, `g_malloc()` is guaranteed to use the system `malloc()`, and `g_free()` just calls the system `free()`.

* https://github.com/GNOME/glib/blob/main/glib/gmem.c#L106
* https://github.com/GNOME/glib/blob/main/glib/gmem.c#L226-L231

So in recent versions, it seems like it's not strictly required, although it is a best practice. To be fair, I can't figure out what the `TRACE(GLIB_MEM_FREE(...))` call is doing. `GLIB_MEM_FREE` is not defined anywhere in the glib source code, so it's something more obscure...

Anyway, Pacemaker only requires glib >= 2.42, so this is moot.